crates/boojum/src/gadgets/u256/mod.rs

```diff
@@ -351,6 +358,64 @@ impl<F: SmallField> UInt256<F> {
         });
         Self::from_le_bytes(cs, bytes)
     }
+
+    /// Finds the result of multiplying `self` by `other` mod `modulo`.
+    pub fn modmul<CS: ConstraintSystem<F>>(
```

Comment:
This function has trouble with generating witnesses. 
2. You check that the module is zero, which is exceptional; you should do this through the constraint is_zero. 
3. You did this wrong first; it creates many inefficiencies because you perform additional range checks on the variables that were already checked. 
4. It looks like you didn't wrap properly carry      
        // let mut carry = (q[0] * p[0] + r[0] - a[0] * b[0]) >> 16;
        // carry = (q[1] * p[0] + q[0] * p[1] + r[1] - a[1] * b[0] - a[0] * b[0] + carry) >> 16;

Reply:
Fixed second point. I am not particularly sure what to change in 3-rd and 4-th, can you please explain wdym?